File: fabric-audit-agent/core/importers/map.js
/**
 * Column mapper: turn an arbitrary CSV table (headers + rows) into the agent's
 * `facts` shape, tolerating the many ways Fabric Capacity Metrics / VertiPaq
 * exports name their columns. Pure: no I/O. Emits a coverage report so callers
 * can show exactly which column fed which field (and what was NOT found).
 *
 * Field contract (must match the detectors):
 *   capacity: { tenant, capacityId, sku, memoryGB, peakCuPct, peakAt, throttleMinutes,
 *               refreshes:[{workspace,dataset,scheduledAt,durationMin,sizeGB}] }
 *   models:  [{ workspace, name, sizeGB, bidirectionalRels, autoDateTime, refreshFailRatePct }]
 *   reports: [{ workspace, name, visuals, mode, slowestVisualMs }]
 */

/** normalize a header for matching: lowercase, keep only a-z 0-9 % */
const norm = (s) => String(s ?? '').toLowerCase().replace(/[^a-z0-9%]/g, '');

/** parse the first number out of a messy cell ("1,234 ms", "87%", "4.2 GB") */
export function num(v) {
  if (v == null) return NaN;
  const m = String(v).replace(/,/g, '').match(/-?\d+(?:\.\d+)?/);
  return m ? parseFloat(m[0]) : NaN;
}

const truthy = (v) => ['true', 'yes', '1', 'enabled', 'on', 'y'].includes(String(v ?? '').trim().toLowerCase());

const round3 = (x) => Math.round(x * 1000) / 1000;

/** Predicates that recognize a column by its normalized header. */
const MATCHERS = {
  sku:          (h) => h === 'sku' || h.includes('skuname') || (h.includes('sku') && !h.includes('skip')),
  capacityName: (h) => h.includes('capacity') && (h.includes('name') || h.includes('id') || h === 'capacity'),
  tenant:       (h) => h.includes('tenant'),
  memoryGB:     (h) => h.includes('capacitymemory') || h === 'memory' || h === 'memorygb' || h === 'ram' || h === 'ramgb',
  throttle:     (h) => h.includes('throttl') || h.includes('overload') || h.includes('interactivedelay'),
  time:         (h) => h.includes('timepoint') || h.includes('timestamp') || h.includes('datetime') || h === 'time' || h === 'date',
  workspace:    (h) => h.includes('workspace'),
  itemName:     (h) => h.includes('itemname') || h === 'item' || h.includes('datasetname') || h === 'dataset' || h.includes('semanticmodel') || (h.includes('model') && h.includes('name')) || h === 'name' || h.includes('reportname'),
  sizeGB:       (h) => h.includes('sizegb') || h.includes('modelsize') || h.includes('datasetsize') || h === 'size' || h.includes('dynamicmemory') || h.includes('totalsize'),
  durationMin:  (h) => h.includes('duration'),
  scheduledAt:  (h) => h.includes('scheduled') || h.includes('starttime') || h === 'start',
  bidi:         (h) => h.includes('bidirection') || h.includes('bidi') || h.includes('bothdirection'),
  autoDate:     (h) => h.includes('autodate') || h.includes('autodatetime'),
  failRate:     (h) => h.includes('failrate') || (h.includes('refresh') && h.includes('fail')) || h.includes('failurerate') || h.includes('errorrate'),
  visuals:      (h) => h.includes('visual') && !h.includes('ms') && !h.includes('slow'),
  mode:         (h) => h === 'mode' || h.includes('storagemode') || h.includes('connectionmode'),
  slowest:      (h) => h.includes('slowest') || h.includes('renderms') || h.includes('rendertime') || (h.includes('visual') && h.includes('ms')),
};

/** find the first header matching `key`, returns the ORIGINAL header text or null */
function find(headers, key) {
  for (const h of headers) if (MATCHERS[key](norm(h))) return h;
  return null;
}

/**
 * Find the capacity *utilization %* column, in priority order. Capacity Metrics
 * exports contain look-alikes that are NOT utilization: "100% in CU(s)" is the
 * baseline (CU-seconds == 100%), "CU % Limit" is the limit line, and
 * "Background %/Interactive %" are component splits. Prefer the real overall-usage
 * column ("Total CU Usage %", "% of base capacity", "Utilization") over those.
 * @param {string[]} headers
 * @returns {string|null}
 */
export function findCuPct(headers) {
  const tagged = headers.map(h => [h, norm(h)]);
  const tiers = [
    ([, h]) => h.includes('totalcuusage') || (h.includes('usage') && h.includes('cu') && h.includes('%')) || h.includes('utiliz'),
    ([, h]) => h.includes('%ofbase') || h.includes('ofbasecapacity'),
    ([, h]) => h.includes('cu') && (h.includes('%') || h.includes('pct') || h.includes('percent'))
      && !h.includes('limit') && !h.includes('100%in') && !h.includes('nonbillable')
      && !h.startsWith('background') && !h.startsWith('interactive') && !h.includes('autoscale'),
  ];
  for (const pred of tiers) { const hit = tagged.find(pred); if (hit) return hit[0]; }
  return null;
}

/** normalize a storage-mode cell to the detector's vocabulary */
function normMode(v) {
  const h = norm(v);
  if (h.includes('direct')) return 'DirectQuery';
  if (h.includes('import')) return 'Import';
  if (h.includes('dual')) return 'Dual';
  if (h.includes('live')) return 'LiveConnection';
  return String(v ?? '').trim();
}

/**
 * Map one table into a partial facts object + coverage report.
 * @param {string[]} headers
 * @param {Record<string,string>[]} rows
 * @returns {{capacity:object|null, models:object[], reports:object[], coverage:Array<{field:string,source:string|null,value:any,note?:string}>}}
 */
export function mapTable(headers, rows = []) {
  const cov = [];
  const cols = {};
  for (const key of Object.keys(MATCHERS)) cols[key] = find(headers, key);
  cols.cuPct = findCuPct(headers);

  const firstNonEmpty = (col) => {
    if (!col) return '';
    for (const r of rows) if (String(r[col] ?? '').trim() !== '') return String(r[col]).trim();
    return '';
  };
  const note = (looked) => `no column found (looked for: ${looked})`;

  // ---- capacity scalars ----
  let capacity = null;
  const hasCapacitySignal = cols.cuPct || cols.throttle || cols.sku || cols.capacityName || cols.memoryGB;
  if (hasCapacitySignal) {
    let peakCuPct = 0, peakAt = '';
    if (cols.cuPct) {
      for (const r of rows) {
        const v = num(r[cols.cuPct]);
        if (Number.isFinite(v) && v > peakCuPct) { peakCuPct = v; peakAt = cols.time ? String(r[cols.time] ?? '').trim() : ''; }
      }
      cov.push({ field: 'peakCuPct', source: cols.cuPct, value: `${peakCuPct}%`, note: peakCuPct > 1000 ? 'that looks like CU-seconds, not a %, double-check the column' : undefined });
    } else cov.push({ field: 'peakCuPct', source: null, value: 0, note: note('Total CU Usage %, utilization, % of base') });

    let throttleMinutes = 0;
    if (cols.throttle) {
      for (const r of rows) { const v = num(r[cols.throttle]); if (Number.isFinite(v)) throttleMinutes += v; }
      cov.push({ field: 'throttleMinutes', source: cols.throttle, value: throttleMinutes });
    } else cov.push({ field: 'throttleMinutes', source: null, value: 0, note: note('throttling, overloaded, rejected') });

    const sku = firstNonEmpty(cols.sku);
    const capacityId = firstNonEmpty(cols.capacityName) || sku || 'unnamed';
    const memoryGB = num(firstNonEmpty(cols.memoryGB));
    cov.push({ field: 'sku', source: cols.sku, value: sku || '(none)' });
    cov.push({ field: 'capacityId', source: cols.capacityName, value: capacityId });

    capacity = {
      tenant: firstNonEmpty(cols.tenant) || 'tenant',
      capacityId,
      sku: sku || '',
      memoryGB: Number.isFinite(memoryGB) ? memoryGB : 0,
      peakCuPct,
      peakAt,
      throttleMinutes,
      refreshes: [],
    };

    // ---- refreshes (one per item row) ----
    if (cols.itemName && (cols.sizeGB || cols.durationMin || cols.scheduledAt)) {
      for (const r of rows) {
        const dataset = String(r[cols.itemName] ?? '').trim();
        if (!dataset) continue;
        capacity.refreshes.push({
          workspace: cols.workspace ? String(r[cols.workspace] ?? '').trim() : '',
          dataset,
          scheduledAt: cols.scheduledAt ? String(r[cols.scheduledAt] ?? '').trim()
                      : cols.time ? String(r[cols.time] ?? '').trim() : '',
          durationMin: cols.durationMin ? (num(r[cols.durationMin]) || 0) : 0,
          sizeGB: cols.sizeGB ? (num(r[cols.sizeGB]) || 0) : 0,
        });
      }
      cov.push({ field: 'capacity.refreshes', source: cols.itemName, value: `${capacity.refreshes.length} row(s)` });
    }
  }

  // ---- models (only if model-shaped columns exist) ----
  const models = [];
  if (cols.itemName && (cols.bidi || cols.autoDate || cols.failRate)) {
    for (const r of rows) {
      const name = String(r[cols.itemName] ?? '').trim();
      if (!name) continue;
      models.push({
        workspace: cols.workspace ? String(r[cols.workspace] ?? '').trim() : '',
        name,
        sizeGB: cols.sizeGB ? (num(r[cols.sizeGB]) || 0) : 0,
        bidirectionalRels: cols.bidi ? (num(r[cols.bidi]) || 0) : 0,
        autoDateTime: cols.autoDate ? truthy(r[cols.autoDate]) : false,
        refreshFailRatePct: cols.failRate ? (num(r[cols.failRate]) || 0) : 0,
      });
    }
    cov.push({ field: 'models', source: cols.itemName, value: `${models.length} row(s)` });
  }

  // ---- reports (only if report-shaped columns exist) ----
  const reports = [];
  if (cols.itemName && (cols.visuals || cols.mode || cols.slowest)) {
    for (const r of rows) {
      const name = String(r[cols.itemName] ?? '').trim();
      if (!name) continue;
      reports.push({
        workspace: cols.workspace ? String(r[cols.workspace] ?? '').trim() : '',
        name,
        visuals: cols.visuals ? (num(r[cols.visuals]) || 0) : 0,
        mode: cols.mode ? normMode(r[cols.mode]) : 'Import',
        slowestVisualMs: cols.slowest ? (num(r[cols.slowest]) || 0) : 0,
      });
    }
    cov.push({ field: 'reports', source: cols.itemName, value: `${reports.length} row(s)` });
  }

  return { capacity, models, reports, coverage: cov };
}

/**
 * Merge several mapped parts into one facts object.
 * Capacity scalars: peakCuPct=max, throttleMinutes=sum, first non-empty for the rest,
 * refreshes concatenated. models/reports concatenated.
 * @param {Array<{capacity:object|null, models:object[], reports:object[]}>} parts
 */
export function mergeFacts(parts) {
  const facts = {};
  const caps = parts.map(p => p?.capacity).filter(Boolean);
  if (caps.length) {
    const capacity = { tenant: 'tenant', capacityId: '', sku: '', memoryGB: 0, peakCuPct: 0, peakAt: '', throttleMinutes: 0, refreshes: [] };
    for (const c of caps) {
      if (c.tenant && capacity.tenant === 'tenant') capacity.tenant = c.tenant;
      capacity.capacityId ||= c.capacityId;
      capacity.sku ||= c.sku;
      capacity.memoryGB ||= c.memoryGB;
      if ((c.peakCuPct || 0) > capacity.peakCuPct) { capacity.peakCuPct = c.peakCuPct; capacity.peakAt = c.peakAt || capacity.peakAt; }
      capacity.throttleMinutes += c.throttleMinutes || 0;
      capacity.refreshes.push(...(c.refreshes || []));
    }
    capacity.capacityId ||= 'unnamed';
    capacity.peakCuPct = round3(capacity.peakCuPct);
    facts.capacity = capacity;
  }
  const models = parts.flatMap(p => p?.models || []);
  if (models.length) facts.models = models;
  const reports = parts.flatMap(p => p?.reports || []);
  if (reports.length) facts.reports = reports;
  return facts;
}

File: fabric-audit-agent-app/fabric_audit_agent/importers/map.py
"""Column mapper. Faithful port of the Node ``core/importers/map.js``.

Turns an arbitrary CSV table into the agent's facts shape, tolerating the many ways
Fabric Capacity Metrics / VertiPaq exports name their columns. Pure: no I/O. Emits a
coverage report so callers can show which column fed which field.
"""
import math
import re

_TRUTHY = {"true", "yes", "1", "enabled", "on", "y"}


def _norm(s):
    return re.sub(r"[^a-z0-9%]", "", str(s if s is not None else "").lower())


def num(v):
    """Parse the first number out of a messy cell ('1,234 ms', '87%', '4.2 GB'); NaN if none."""
    if v is None:
        return float("nan")
    m = re.search(r"-?\d+(?:\.\d+)?", str(v).replace(",", ""))
    return float(m.group(0)) if m else float("nan")


def _num0(v):
    """num(v) coerced to 0 when non-finite — mirrors JS `num(v) || 0` (NaN is falsy in JS)."""
    n = num(v)
    return n if math.isfinite(n) else 0


def _truthy(v):
    return str(v if v is not None else "").strip().lower() in _TRUTHY


def _round3(x):
    return math.floor(x * 1000 + 0.5) / 1000   # JS Math.round (half-up); values >= 0


def _fmt(x):
    return str(int(x)) if x == int(x) else str(x)


_MATCHERS = {
    "sku": lambda h: h == "sku" or "skuname" in h or ("sku" in h and "skip" not in h),
    "capacityName": lambda h: "capacity" in h and ("name" in h or "id" in h or h == "capacity"),
    "tenant": lambda h: "tenant" in h,
    "memoryGB": lambda h: "capacitymemory" in h or h in ("memory", "memorygb", "ram", "ramgb"),
    "throttle": lambda h: "throttl" in h or "overload" in h or "interactivedelay" in h,
    "time": lambda h: "timepoint" in h or "timestamp" in h or "datetime" in h or h in ("time", "date"),
    "workspace": lambda h: "workspace" in h,
    "itemName": lambda h: "itemname" in h or h == "item" or "datasetname" in h or h == "dataset" or "semanticmodel" in h or ("model" in h and "name" in h) or h == "name" or "reportname" in h,
    "sizeGB": lambda h: "sizegb" in h or "modelsize" in h or "datasetsize" in h or h == "size" or "dynamicmemory" in h or "totalsize" in h,
    "durationMin": lambda h: "duration" in h,
    "scheduledAt": lambda h: "scheduled" in h or "starttime" in h or h == "start",
    "bidi": lambda h: "bidirection" in h or "bidi" in h or "bothdirection" in h,
    "autoDate": lambda h: "autodate" in h or "autodatetime" in h,
    "failRate": lambda h: "failrate" in h or ("refresh" in h and "fail" in h) or "failurerate" in h or "errorrate" in h,
    "visuals": lambda h: "visual" in h and "ms" not in h and "slow" not in h,
    "mode": lambda h: h == "mode" or "storagemode" in h or "connectionmode" in h,
    "slowest": lambda h: "slowest" in h or "renderms" in h or "rendertime" in h or ("visual" in h and "ms" in h),
}


def _find(headers, key):
    for h in headers:
        if _MATCHERS[key](_norm(h)):
            return h
    return None


def find_cu_pct(headers):
    """Find the capacity utilization % column, preferring true overall-usage over
    look-alikes ('100% in CU(s)' baseline, 'CU % Limit', 'Background %/Interactive %')."""
    tagged = [(h, _norm(h)) for h in headers]
    tiers = [
        lambda t: "totalcuusage" in t[1] or ("usage" in t[1] and "cu" in t[1] and "%" in t[1]) or "utiliz" in t[1],
        lambda t: "%ofbase" in t[1] or "ofbasecapacity" in t[1],
        lambda t: "cu" in t[1] and ("%" in t[1] or "pct" in t[1] or "percent" in t[1])
        and "limit" not in t[1] and "100%in" not in t[1] and "nonbillable" not in t[1]
        and not t[1].startswith("background") and not t[1].startswith("interactive") and "autoscale" not in t[1],
    ]
    for pred in tiers:
        for t in tagged:
            if pred(t):
                return t[0]
    return None


def _norm_mode(v):
    h = _norm(v)
    if "direct" in h:
        return "DirectQuery"
    if "import" in h:
        return "Import"
    if "dual" in h:
        return "Dual"
    if "live" in h:
        return "LiveConnection"
    return str(v if v is not None else "").strip()


def _cell(r, col):
    return str(r.get(col) or "").strip() if col else ""


def map_table(headers, rows=None):
    rows = rows or []
    cov = []
    cols = {key: _find(headers, key) for key in _MATCHERS}
    cols["cuPct"] = find_cu_pct(headers)

    def first_non_empty(col):
        if not col:
            return ""
        for r in rows:
            v = str(r.get(col) or "").strip()
            if v != "":
                return v
        return ""

    def note(looked):
        return f"no column found (looked for: {looked})"

    capacity = None
    has_capacity_signal = cols["cuPct"] or cols["throttle"] or cols["sku"] or cols["capacityName"] or cols["memoryGB"]
    if has_capacity_signal:
        peak_cu_pct = 0
        peak_at = ""
        if cols["cuPct"]:
            for r in rows:
                v = num(r.get(cols["cuPct"]))
                if math.isfinite(v) and v > peak_cu_pct:
                    peak_cu_pct = v
                    peak_at = _cell(r, cols["time"]) if cols["time"] else ""
            entry = {"field": "peakCuPct", "source": cols["cuPct"], "value": f"{_fmt(peak_cu_pct)}%"}
            if peak_cu_pct > 1000:
                entry["note"] = "that looks like CU-seconds, not a %, double-check the column"
            cov.append(entry)
        else:
            cov.append({"field": "peakCuPct", "source": None, "value": 0, "note": note("Total CU Usage %, utilization, % of base")})

        throttle_minutes = 0
        if cols["throttle"]:
            for r in rows:
                v = num(r.get(cols["throttle"]))
                if math.isfinite(v):
                    throttle_minutes += v
            cov.append({"field": "throttleMinutes", "source": cols["throttle"], "value": throttle_minutes})
        else:
            cov.append({"field": "throttleMinutes", "source": None, "value": 0, "note": note("throttling, overloaded, rejected")})

        sku = first_non_empty(cols["sku"])
        capacity_id = first_non_empty(cols["capacityName"]) or sku or "unnamed"
        memory_gb = num(first_non_empty(cols["memoryGB"]))
        cov.append({"field": "sku", "source": cols["sku"], "value": sku or "(none)"})
        cov.append({"field": "capacityId", "source": cols["capacityName"], "value": capacity_id})

        capacity = {
            "tenant": first_non_empty(cols["tenant"]) or "tenant",
            "capacityId": capacity_id,
            "sku": sku or "",
            "memoryGB": memory_gb if math.isfinite(memory_gb) else 0,
            "peakCuPct": peak_cu_pct,
            "peakAt": peak_at,
            "throttleMinutes": throttle_minutes,
            "refreshes": [],
        }

        if cols["itemName"] and (cols["sizeGB"] or cols["durationMin"] or cols["scheduledAt"]):
            for r in rows:
                dataset = _cell(r, cols["itemName"])
                if not dataset:
                    continue
                if cols["scheduledAt"]:
                    sched = _cell(r, cols["scheduledAt"])
                elif cols["time"]:
                    sched = _cell(r, cols["time"])
                else:
                    sched = ""
                capacity["refreshes"].append({
                    "workspace": _cell(r, cols["workspace"]),
                    "dataset": dataset,
                    "scheduledAt": sched,
                    "durationMin": _num0(r.get(cols["durationMin"])) if cols["durationMin"] else 0,
                    "sizeGB": _num0(r.get(cols["sizeGB"])) if cols["sizeGB"] else 0,
                })
            cov.append({"field": "capacity.refreshes", "source": cols["itemName"], "value": f"{len(capacity['refreshes'])} row(s)"})

    models = []
    if cols["itemName"] and (cols["bidi"] or cols["autoDate"] or cols["failRate"]):
        for r in rows:
            name = _cell(r, cols["itemName"])
            if not name:
                continue
            models.append({
                "workspace": _cell(r, cols["workspace"]),
                "name": name,
                "sizeGB": _num0(r.get(cols["sizeGB"])) if cols["sizeGB"] else 0,
                "bidirectionalRels": _num0(r.get(cols["bidi"])) if cols["bidi"] else 0,
                "autoDateTime": _truthy(r.get(cols["autoDate"])) if cols["autoDate"] else False,
                "refreshFailRatePct": _num0(r.get(cols["failRate"])) if cols["failRate"] else 0,
            })
        cov.append({"field": "models", "source": cols["itemName"], "value": f"{len(models)} row(s)"})

    reports = []
    if cols["itemName"] and (cols["visuals"] or cols["mode"] or cols["slowest"]):
        for r in rows:
            name = _cell(r, cols["itemName"])
            if not name:
                continue
            reports.append({
                "workspace": _cell(r, cols["workspace"]),
                "name": name,
                "visuals": _num0(r.get(cols["visuals"])) if cols["visuals"] else 0,
                "mode": _norm_mode(r.get(cols["mode"])) if cols["mode"] else "Import",
                "slowestVisualMs": _num0(r.get(cols["slowest"])) if cols["slowest"] else 0,
            })
        cov.append({"field": "reports", "source": cols["itemName"], "value": f"{len(reports)} row(s)"})

    return {"capacity": capacity, "models": models, "reports": reports, "coverage": cov}


def merge_facts(parts):
    facts = {}
    caps = [p.get("capacity") for p in parts if p and p.get("capacity")]
    if caps:
        capacity = {"tenant": "tenant", "capacityId": "", "sku": "", "memoryGB": 0, "peakCuPct": 0, "peakAt": "", "throttleMinutes": 0, "refreshes": []}
        for c in caps:
            if c.get("tenant") and capacity["tenant"] == "tenant":
                capacity["tenant"] = c["tenant"]
            capacity["capacityId"] = capacity["capacityId"] or c.get("capacityId")
            capacity["sku"] = capacity["sku"] or c.get("sku")
            capacity["memoryGB"] = capacity["memoryGB"] or c.get("memoryGB")
            if (c.get("peakCuPct") or 0) > capacity["peakCuPct"]:
                capacity["peakCuPct"] = c.get("peakCuPct")
                capacity["peakAt"] = c.get("peakAt") or capacity["peakAt"]
            capacity["throttleMinutes"] += c.get("throttleMinutes") or 0
            capacity["refreshes"].extend(c.get("refreshes") or [])
        capacity["capacityId"] = capacity["capacityId"] or "unnamed"
        capacity["peakCuPct"] = _round3(capacity["peakCuPct"])
        facts["capacity"] = capacity
    models = [m for p in parts for m in (p.get("models") or [])]
    if models:
        facts["models"] = models
    reports = [r for p in parts for r in (p.get("reports") or [])]
    if reports:
        facts["reports"] = reports
    return facts
